File: packages/health/lib/src/data_types.dart
part of health;

/// List of all available data types.
enum HealthDataType {
  ACTIVE_ENERGY_BURNED,
  AUDIOGRAM,
  BASAL_ENERGY_BURNED,
  BLOOD_GLUCOSE,
  BLOOD_OXYGEN,
  BLOOD_PRESSURE_DIASTOLIC,
  BLOOD_PRESSURE_SYSTOLIC,
  BODY_FAT_PERCENTAGE,
  BODY_MASS_INDEX,
  BODY_TEMPERATURE,
  DIETARY_CARBS_CONSUMED,
  DIETARY_ENERGY_CONSUMED,
  DIETARY_FATS_CONSUMED,
  DIETARY_PROTEIN_CONSUMED,
  FORCED_EXPIRATORY_VOLUME,
  HEART_RATE,
  HEART_RATE_VARIABILITY_SDNN,
  HEIGHT,
  RESTING_HEART_RATE,
  STEPS,
  WAIST_CIRCUMFERENCE,
  WALKING_HEART_RATE,
  WEIGHT,
  DISTANCE_WALKING_RUNNING,
  FLIGHTS_CLIMBED,
  MOVE_MINUTES,
  DISTANCE_DELTA,
  MINDFULNESS,
  WATER,
  SLEEP_IN_BED,
  SLEEP_ASLEEP,
  SLEEP_AWAKE,
  EXERCISE_TIME,
  WORKOUT,
  HEADACHE_NOT_PRESENT,
  HEADACHE_MILD,
  HEADACHE_MODERATE,
  HEADACHE_SEVERE,
  HEADACHE_UNSPECIFIED,

  // Heart Rate events (specific to Apple Watch)
  HIGH_HEART_RATE_EVENT,
  LOW_HEART_RATE_EVENT,
  IRREGULAR_HEART_RATE_EVENT,
  ELECTRODERMAL_ACTIVITY,
  ELECTROCARDIOGRAM,
}

enum HealthDataAccess {
  READ,
  WRITE,
  READ_WRITE,
}

/// List of data types available on iOS
const List<HealthDataType> _dataTypeKeysIOS = [
  HealthDataType.ACTIVE_ENERGY_BURNED,
  HealthDataType.AUDIOGRAM,
  HealthDataType.BASAL_ENERGY_BURNED,
  HealthDataType.BLOOD_GLUCOSE,
  HealthDataType.BLOOD_OXYGEN,
  HealthDataType.BLOOD_PRESSURE_DIASTOLIC,
  HealthDataType.BLOOD_PRESSURE_SYSTOLIC,
  HealthDataType.BODY_FAT_PERCENTAGE,
  HealthDataType.BODY_MASS_INDEX,
  HealthDataType.BODY_TEMPERATURE,
  HealthDataType.DIETARY_CARBS_CONSUMED,
  HealthDataType.DIETARY_ENERGY_CONSUMED,
  HealthDataType.DIETARY_FATS_CONSUMED,
  HealthDataType.DIETARY_PROTEIN_CONSUMED,
  HealthDataType.ELECTRODERMAL_ACTIVITY,
  HealthDataType.FORCED_EXPIRATORY_VOLUME,
  HealthDataType.HEART_RATE,
  HealthDataType.HEART_RATE_VARIABILITY_SDNN,
  HealthDataType.HEIGHT,
  HealthDataType.HIGH_HEART_RATE_EVENT,
  HealthDataType.IRREGULAR_HEART_RATE_EVENT,
  HealthDataType.LOW_HEART_RATE_EVENT,
  HealthDataType.RESTING_HEART_RATE,
  HealthDataType.STEPS,
  HealthDataType.WAIST_CIRCUMFERENCE,
  HealthDataType.WALKING_HEART_RATE,
  HealthDataType.WEIGHT,
  HealthDataType.FLIGHTS_CLIMBED,
  HealthDataType.DISTANCE_WALKING_RUNNING,
  HealthDataType.MINDFULNESS,
  HealthDataType.SLEEP_IN_BED,
  HealthDataType.SLEEP_AWAKE,
  HealthDataType.SLEEP_ASLEEP,
  HealthDataType.WATER,
  HealthDataType.EXERCISE_TIME,
  HealthDataType.WORKOUT,
  HealthDataType.HEADACHE_NOT_PRESENT,
  HealthDataType.HEADACHE_MILD,
  HealthDataType.HEADACHE_MODERATE,
  HealthDataType.HEADACHE_SEVERE,
  HealthDataType.HEADACHE_UNSPECIFIED,
  HealthDataType.ELECTROCARDIOGRAM,
];

/// List of data types available on Android
const List<HealthDataType> _dataTypeKeysAndroid = [
  HealthDataType.ACTIVE_ENERGY_BURNED,
  HealthDataType.BLOOD_GLUCOSE,
  HealthDataType.BLOOD_OXYGEN,
  HealthDataType.BLOOD_PRESSURE_DIASTOLIC,
  HealthDataType.BLOOD_PRESSURE_SYSTOLIC,
  HealthDataType.BODY_FAT_PERCENTAGE,
  HealthDataType.BODY_MASS_INDEX,
  HealthDataType.BODY_TEMPERATURE,
  HealthDataType.HEART_RATE,
  HealthDataType.HEIGHT,
  HealthDataType.STEPS,
  HealthDataType.WEIGHT,
  HealthDataType.MOVE_MINUTES,
  HealthDataType.DISTANCE_DELTA,
  HealthDataType.SLEEP_AWAKE,
  HealthDataType.SLEEP_ASLEEP,
  HealthDataType.SLEEP_IN_BED,
  HealthDataType.WATER,
  HealthDataType.WORKOUT,
];

/// Maps a [HealthDataType] to a [HealthDataUnit].
const Map<HealthDataType, HealthDataUnit> _dataTypeToUnit = {
  HealthDataType.ACTIVE_ENERGY_BURNED: HealthDataUnit.KILOCALORIE,
  HealthDataType.AUDIOGRAM: HealthDataUnit.DECIBEL_HEARING_LEVEL,
  HealthDataType.BASAL_ENERGY_BURNED: HealthDataUnit.KILOCALORIE,
  HealthDataType.BLOOD_GLUCOSE: HealthDataUnit.MILLIGRAM_PER_DECILITER,
  HealthDataType.BLOOD_OXYGEN: HealthDataUnit.PERCENT,
  HealthDataType.BLOOD_PRESSURE_DIASTOLIC: HealthDataUnit.MILLIMETER_OF_MERCURY,
  HealthDataType.BLOOD_PRESSURE_SYSTOLIC: HealthDataUnit.MILLIMETER_OF_MERCURY,
  HealthDataType.BODY_FAT_PERCENTAGE: HealthDataUnit.PERCENT,
  HealthDataType.BODY_MASS_INDEX: HealthDataUnit.NO_UNIT,
  HealthDataType.BODY_TEMPERATURE: HealthDataUnit.DEGREE_CELSIUS,
  HealthDataType.DIETARY_CARBS_CONSUMED: HealthDataUnit.GRAM,
  HealthDataType.DIETARY_ENERGY_CONSUMED: HealthDataUnit.KILOCALORIE,
  HealthDataType.DIETARY_FATS_CONSUMED: HealthDataUnit.GRAM,
  HealthDataType.DIETARY_PROTEIN_CONSUMED: HealthDataUnit.GRAM,
  HealthDataType.ELECTRODERMAL_ACTIVITY: HealthDataUnit.SIEMEN,
  HealthDataType.FORCED_EXPIRATORY_VOLUME: HealthDataUnit.LITER,
  HealthDataType.HEART_RATE: HealthDataUnit.BEATS_PER_MINUTE,
  HealthDataType.HEIGHT: HealthDataUnit.METER,
  HealthDataType.RESTING_HEART_RATE: HealthDataUnit.BEATS_PER_MINUTE,
  HealthDataType.STEPS: HealthDataUnit.COUNT,
  HealthDataType.WAIST_CIRCUMFERENCE: HealthDataUnit.METER,
  HealthDataType.WALKING_HEART_RATE: HealthDataUnit.BEATS_PER_MINUTE,
  HealthDataType.WEIGHT: HealthDataUnit.KILOGRAM,
  HealthDataType.DISTANCE_WALKING_RUNNING: HealthDataUnit.METER,
  HealthDataType.FLIGHTS_CLIMBED: HealthDataUnit.COUNT,
  HealthDataType.MOVE_MINUTES: HealthDataUnit.MINUTE,
  HealthDataType.DISTANCE_DELTA: HealthDataUnit.METER,

  HealthDataType.WATER: HealthDataUnit.LITER,
  HealthDataType.SLEEP_IN_BED: HealthDataUnit.MINUTE,
  HealthDataType.SLEEP_ASLEEP: HealthDataUnit.MINUTE,
  HealthDataType.SLEEP_AWAKE: HealthDataUnit.MINUTE,
  HealthDataType.MINDFULNESS: HealthDataUnit.MINUTE,
  HealthDataType.EXERCISE_TIME: HealthDataUnit.MINUTE,
  HealthDataType.WORKOUT: HealthDataUnit.NO_UNIT,

  HealthDataType.HEADACHE_NOT_PRESENT: HealthDataUnit.MINUTE,
  HealthDataType.HEADACHE_MILD: HealthDataUnit.MINUTE,
  HealthDataType.HEADACHE_MODERATE: HealthDataUnit.MINUTE,
  HealthDataType.HEADACHE_SEVERE: HealthDataUnit.MINUTE,
  HealthDataType.HEADACHE_UNSPECIFIED: HealthDataUnit.MINUTE,

  // Heart Rate events (specific to Apple Watch)
  HealthDataType.HIGH_HEART_RATE_EVENT: HealthDataUnit.NO_UNIT,
  HealthDataType.LOW_HEART_RATE_EVENT: HealthDataUnit.NO_UNIT,
  HealthDataType.IRREGULAR_HEART_RATE_EVENT: HealthDataUnit.NO_UNIT,
  HealthDataType.HEART_RATE_VARIABILITY_SDNN: HealthDataUnit.MILLISECOND,
  HealthDataType.ELECTROCARDIOGRAM: HealthDataUnit.VOLT,
};

const PlatformTypeJsonValue = {
  PlatformType.IOS: 'ios',
  PlatformType.ANDROID: 'android',
};

/// List of all [HealthDataUnit]s.
enum HealthDataUnit {
  // Mass units
  GRAM,
  KILOGRAM,
  OUNCE,
  POUND,
  STONE,
  // MOLE_UNIT_WITH_MOLAR_MASS, // requires molar mass input - not supported yet
  // MOLE_UNIT_WITH_PREFIX_MOLAR_MASS, // requires molar mass & prefix input - not supported yet

  // Length units
  METER,
  INCH,
  FOOT,
  YARD,
  MILE,

  // Volume units
  LITER,
  MILLILITER,
  FLUID_OUNCE_US,
  FLUID_OUNCE_IMPERIAL,
  CUP_US,
  CUP_IMPERIAL,
  PINT_US,
  PINT_IMPERIAL,

  // Pressure units
  PASCAL,
  MILLIMETER_OF_MERCURY,
  INCHES_OF_MERCURY,
  CENTIMETER_OF_WATER,
  ATMOSPHERE,
  DECIBEL_A_WEIGHTED_SOUND_PRESSURE_LEVEL,

  // Time units
  SECOND,
  MILLISECOND,
  MINUTE,
  HOUR,
  DAY,

  // Energy units
  JOULE,
  KILOCALORIE,
  LARGE_CALORIE,
  SMALL_CALORIE,

  // Temperature units
  DEGREE_CELSIUS,
  DEGREE_FAHRENHEIT,
  KELVIN,

  // Hearing units
  DECIBEL_HEARING_LEVEL,

  // Frequency units
  HERTZ,

  // Electrical conductance units
  SIEMEN,

  // Potential units
  VOLT,

  // Pharmacology units
  INTERNATIONAL_UNIT,

  // Scalar units
  COUNT,
  PERCENT,

  // Other units
  BEATS_PER_MINUTE,
  MILLIGRAM_PER_DECILITER,
  UNKNOWN_UNIT,
  NO_UNIT,
}

/// List of [HealthWorkoutActivityType]s.
/// Commented for which platform they are supported
enum HealthWorkoutActivityType {
  // Both
  ARCHERY,
  BADMINTON,
  BASEBALL,
  BASKETBALL,
  BIKING, // This also entails the iOS version where it is called CYCLING
  BOXING,
  CRICKET,
  CURLING,
  ELLIPTICAL,
  FENCING,
  AMERICAN_FOOTBALL,
  AUSTRALIAN_FOOTBALL,
  SOCCER,
  GOLF,
  GYMNASTICS,
  HANDBALL,
  HIGH_INTENSITY_INTERVAL_TRAINING,
  HIKING,
  HOCKEY,
  SKATING,
  JUMP_ROPE,
  KICKBOXING,
  MARTIAL_ARTS,
  PILATES,
  RACQUETBALL,
  ROWING,
  RUGBY,
  RUNNING,
  SAILING,
  CROSS_COUNTRY_SKIING,
  DOWNHILL_SKIING,
  SNOWBOARDING,
  SOFTBALL,
  SQUASH,
  STAIR_CLIMBING,
  SWIMMING,
  TABLE_TENNIS,
  TENNIS,
  VOLLEYBALL,
  WALKING,
  WATER_POLO,
  YOGA,

  // iOS only
  BOWLING,
  CROSS_TRAINING,
  TRACK_AND_FIELD,
  DISC_SPORTS,
  LACROSSE,
  PREPARATION_AND_RECOVERY,
  FLEXIBILITY,
  COOLDOWN,
  WHEELCHAIR_WALK_PACE,
  WHEELCHAIR_RUN_PACE,
  HAND_CYCLING,
  CORE_TRAINING,
  FUNCTIONAL_STRENGTH_TRAINING,
  TRADITIONAL_STRENGTH_TRAINING,
  MIXED_CARDIO,
  STAIRS,
  STEP_TRAINING,
  FITNESS_GAMING,
  BARRE,
  CARDIO_DANCE,
  SOCIAL_DANCE,
  MIND_AND_BODY,
  PICKLEBALL,
  CLIMBING,
  EQUESTRIAN_SPORTS,
  FISHING,
  HUNTING,
  PLAY,
  SNOW_SPORTS,
  PADDLE_SPORTS,
  SURFING_SPORTS,
  WATER_FITNESS,
  WATER_SPORTS,
  TAI_CHI,
  WRESTLING,

  // Android only
  AEROBICS,
  BIATHLON,
  BIKING_HAND,
  BIKING_MOUNTAIN,
  BIKING_ROAD,
  BIKING_SPINNING,
  BIKING_STATIONARY,
  BIKING_UTILITY,
  CALISTHENICS,
  CIRCUIT_TRAINING,
  CROSS_FIT,
  DANCING,
  DIVING,
  ELEVATOR,
  ERGOMETER,
  ESCALATOR,
  FRISBEE_DISC,
  GARDENING,
  GUIDED_BREATHING,
  HORSEBACK_RIDING,
  HOUSEWORK,
  INTERVAL_TRAINING,
  IN_VEHICLE,
  ICE_SKATING,
  KAYAKING,
  KETTLEBELL_TRAINING,
  KICK_SCOOTER,
  KITE_SURFING,
  MEDITATION,
  MIXED_MARTIAL_ARTS,
  P90X,
  PARAGLIDING,
  POLO,
  ROCK_CLIMBING, // on iOS this is the same as CLIMBING
  ROWING_MACHINE,
  RUNNING_JOGGING, // on iOS this is the same as RUNNING
  RUNNING_SAND, // on iOS this is the same as RUNNING
  RUNNING_TREADMILL, // on iOS this is the same as RUNNING
  SCUBA_DIVING,
  SKATING_CROSS, // on iOS this is the same as SKATING
  SKATING_INDOOR, // on iOS this is the same as SKATING
  SKATING_INLINE, // on iOS this is the same as SKATING
  SKIING,
  SKIING_BACK_COUNTRY,
  SKIING_KITE,
  SKIING_ROLLER,
  SLEDDING,
  SNOWMOBILE,
  SNOWSHOEING,
  STAIR_CLIMBING_MACHINE,
  STANDUP_PADDLEBOARDING,
  STILL,
  STRENGTH_TRAINING,
  SURFING,
  SWIMMING_OPEN_WATER,
  SWIMMING_POOL,
  TEAM_SPORTS,
  TILTING,
  VOLLEYBALL_BEACH,
  VOLLEYBALL_INDOOR,
  WAKEBOARDING,
  WALKING_FITNESS,
  WALKING_NORDIC,
  WALKING_STROLLER,
  WALKING_TREADMILL,
  WEIGHTLIFTING,
  WHEELCHAIR,
  WINDSURFING,
  ZUMBA,

  //
  OTHER,
}

enum ElectrocardiogramClassification {
  NOT_SET,
  SINUS_RHYTHM,
  ATRIAL_FIBRILLATION,
  INCONCLUSIVE_LOW_HEART_RATE,
  INCONCLUSIVE_HIGH_HEART_RATE,
  INCONCLUSIVE_POOR_READING,
  INCONCLUSIVE_OTHER,
  UNRECOGNIZED,
}

extension ElectrocardiogramClassificationValue
    on ElectrocardiogramClassification {
  int get value {
    switch (this) {
      case ElectrocardiogramClassification.NOT_SET:
        return 0;
      case ElectrocardiogramClassification.SINUS_RHYTHM:
        return 1;
      case ElectrocardiogramClassification.ATRIAL_FIBRILLATION:
        return 2;
      case ElectrocardiogramClassification.INCONCLUSIVE_LOW_HEART_RATE:
        return 3;
      case ElectrocardiogramClassification.INCONCLUSIVE_HIGH_HEART_RATE:
        return 4;
      case ElectrocardiogramClassification.INCONCLUSIVE_POOR_READING:
        return 5;
      case ElectrocardiogramClassification.INCONCLUSIVE_OTHER:
        return 6;
      case ElectrocardiogramClassification.UNRECOGNIZED:
        return 100;
    }
  }
}

File: packages/health/ios/Classes/DataTypes.swift

import HealthKit

enum HealthWorkoutActivityType: String, CaseIterable {
    // Both
    case ARCHERY
    case BADMINTON
    case BASEBALL
    case BASKETBALL
    case BIKING  // This also entails the iOS version where it is called CYCLING
    case BOXING
    case CRICKET
    case CURLING
    case ELLIPTICAL
    case FENCING
    case AMERICAN_FOOTBALL
    case AUSTRALIAN_FOOTBALL
    case SOCCER
    case GOLF
    case GYMNASTICS
    case HANDBALL
    case HIGH_INTENSITY_INTERVAL_TRAINING
    case HIKING
    case HOCKEY
    case SKATING
    case JUMP_ROPE
    case KICKBOXING
    case MARTIAL_ARTS
    case PILATES
    case RACQUETBALL
    case ROWING
    case RUGBY
    case RUNNING
    case SAILING
    case CROSS_COUNTRY_SKIING
    case DOWNHILL_SKIING
    case SNOWBOARDING
    case SOFTBALL
    case SQUASH
    case STAIR_CLIMBING
    case SWIMMING
    case TABLE_TENNIS
    case TENNIS
    case VOLLEYBALL
    case WALKING
    case WATER_POLO
    case YOGA

    // iOS only
    case BOWLING
    case CROSS_TRAINING
    case TRACK_AND_FIELD
    case DISC_SPORTS
    case LACROSSE
    case PREPARATION_AND_RECOVERY
    case FLEXIBILITY
    case COOLDOWN
    case WHEELCHAIR_WALK_PACE
    case WHEELCHAIR_RUN_PACE
    case HAND_CYCLING
    case CORE_TRAINING
    case FUNCTIONAL_STRENGTH_TRAINING
    case TRADITIONAL_STRENGTH_TRAINING
    case MIXED_CARDIO
    case STAIRS
    case STEP_TRAINING
    case FITNESS_GAMING
    case BARRE
    case CARDIO_DANCE
    case SOCIAL_DANCE
    case MIND_AND_BODY
    case PICKLEBALL
    case CLIMBING
    case EQUESTRIAN_SPORTS
    case FISHING
    case HUNTING
    case PLAY
    case SNOW_SPORTS
    case PADDLE_SPORTS
    case SURFING_SPORTS
    case WATER_FITNESS
    case WATER_SPORTS
    case TAI_CHI
    case WRESTLING

    // Android only
    case AEROBICS
    case BIATHLON
    case CALISTHENICS
    case CIRCUIT_TRAINING
    case CROSS_FIT
    case DANCING
    case DIVING
    case ELEVATOR
    case ERGOMETER
    case ESCALATOR
    case FRISBEE_DISC
    case GARDENING
    case GUIDED_BREATHING
    case HORSEBACK_RIDING
    case HOUSEWORK
    case INTERVAL_TRAINING
    case IN_VEHICLE
    case KAYAKING
    case KETTLEBELL_TRAINING
    case KICK_SCOOTER
    case KITE_SURFING
    case MEDITATION
    case MIXED_MARTIAL_ARTS
    case P90X
    case PARAGLIDING
    case POLO
    case ROCK_CLIMBING // on iOS this is the same as CLIMBING
    case RUNNING_JOGGING // on iOS this is the same as RUNNING
    case RUNNING_SAND // on iOS this is the same as RUNNING
    case RUNNING_TREADMILL // on iOS this is the same as RUNNING
    case SCUBA_DIVING
    case SKATING_CROSS // on iOS this is the same as SKATING
    case SKATING_INDOOR // on iOS this is the same as SKATING
    case SKATING_INLINE // on iOS this is the same as SKATING
    case SKIING_BACK_COUNTRY
    case SKIING_KITE
    case SKIING_ROLLER
    case SLEDDING
    case STAIR_CLIMBING_MACHINE
    case STANDUP_PADDLEBOARDING
    case STILL
    case STRENGTH_TRAINING
    case SURFING
    case SWIMMING_OPEN_WATER
    case SWIMMING_POOL
    case TEAM_SPORTS
    case TILTING
    case VOLLEYBALL_BEACH
    case VOLLEYBALL_INDOOR
    case WAKEBOARDING
    case WALKING_FITNESS
    case WALKING_NORDIC
    case WALKING_STROLLER
    case WALKING_TREADMILL
    case WEIGHTLIFTING
    case WHEELCHAIR
    case WINDSURFING
    case ZUMBA

    //
    case OTHER

    static func withLabel(_ label: String) -> HealthWorkoutActivityType? {
      return self.allCases.first{ "\($0)" == label }
    }
}


class HealthToHealthkitActivityMapping {
    var healthActivityTypeToHealthkit: [HealthWorkoutActivityType: HKWorkoutActivityType] = [:]
    var healthkitTypeToHealth: [HKWorkoutActivityType: HealthWorkoutActivityType] = [:]

    init() {
        healthActivityTypeToHealthkit[.ARCHERY] = .archery
        healthActivityTypeToHealthkit[.BOWLING] = .bowling
        healthActivityTypeToHealthkit[.FENCING] = .fencing
        healthActivityTypeToHealthkit[.GYMNASTICS] = .gymnastics
        healthActivityTypeToHealthkit[.TRACK_AND_FIELD] = .trackAndField
        healthActivityTypeToHealthkit[.AMERICAN_FOOTBALL] = .americanFootball
        healthActivityTypeToHealthkit[.AUSTRALIAN_FOOTBALL] = .australianFootball
        healthActivityTypeToHealthkit[.BASEBALL] = .baseball
        healthActivityTypeToHealthkit[.BASKETBALL] = .basketball
        healthActivityTypeToHealthkit[.CRICKET] = .cricket
        healthActivityTypeToHealthkit[.DISC_SPORTS] = .discSports
        healthActivityTypeToHealthkit[.HANDBALL] = .handball
        healthActivityTypeToHealthkit[.HOCKEY] = .hockey
        healthActivityTypeToHealthkit[.LACROSSE] = .lacrosse
        healthActivityTypeToHealthkit[.RUGBY] = .rugby
        healthActivityTypeToHealthkit[.SOCCER] = .soccer
        healthActivityTypeToHealthkit[.SOFTBALL] = .softball
        healthActivityTypeToHealthkit[.VOLLEYBALL] = .volleyball
        healthActivityTypeToHealthkit[.PREPARATION_AND_RECOVERY] = .preparationAndRecovery
        healthActivityTypeToHealthkit[.FLEXIBILITY] = .flexibility
        healthActivityTypeToHealthkit[.WALKING] = .walking
        healthActivityTypeToHealthkit[.RUNNING] = .running
        healthActivityTypeToHealthkit[.RUNNING_JOGGING] = .running // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.RUNNING_SAND] = .running // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.RUNNING_TREADMILL] = .running // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.WHEELCHAIR_WALK_PACE] = .wheelchairWalkPace
        healthActivityTypeToHealthkit[.WHEELCHAIR_RUN_PACE] = .wheelchairRunPace
        healthActivityTypeToHealthkit[.BIKING] = .cycling
        healthActivityTypeToHealthkit[.HAND_CYCLING] = .handCycling
        healthActivityTypeToHealthkit[.CORE_TRAINING] = .coreTraining
        healthActivityTypeToHealthkit[.ELLIPTICAL] = .elliptical
        healthActivityTypeToHealthkit[.FUNCTIONAL_STRENGTH_TRAINING] = .functionalStrengthTraining
        healthActivityTypeToHealthkit[.TRADITIONAL_STRENGTH_TRAINING] = .traditionalStrengthTraining
        healthActivityTypeToHealthkit[.CROSS_TRAINING] = .crossTraining
        healthActivityTypeToHealthkit[.MIXED_CARDIO] = .mixedCardio
        healthActivityTypeToHealthkit[.HIGH_INTENSITY_INTERVAL_TRAINING] = .highIntensityIntervalTraining
        healthActivityTypeToHealthkit[.JUMP_ROPE] = .jumpRope
        healthActivityTypeToHealthkit[.STAIR_CLIMBING] = .stairClimbing
        healthActivityTypeToHealthkit[.STAIRS] = .stairs
        healthActivityTypeToHealthkit[.STEP_TRAINING] = .stepTraining
        healthActivityTypeToHealthkit[.FITNESS_GAMING] = .fitnessGaming
        healthActivityTypeToHealthkit[.BARRE] = .barre
        healthActivityTypeToHealthkit[.YOGA] = .yoga
        healthActivityTypeToHealthkit[.MIND_AND_BODY] = .mindAndBody
        healthActivityTypeToHealthkit[.PILATES] = .pilates
        healthActivityTypeToHealthkit[.BADMINTON] = .badminton
        healthActivityTypeToHealthkit[.RACQUETBALL] = .racquetball
        healthActivityTypeToHealthkit[.SQUASH] = .squash
        healthActivityTypeToHealthkit[.TABLE_TENNIS] = .tableTennis
        healthActivityTypeToHealthkit[.TENNIS] = .tennis
        healthActivityTypeToHealthkit[.CLIMBING] = .climbing
        healthActivityTypeToHealthkit[.ROCK_CLIMBING] = .climbing // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.EQUESTRIAN_SPORTS] = .equestrianSports
        healthActivityTypeToHealthkit[.FISHING] = .fishing
        healthActivityTypeToHealthkit[.GOLF] = .golf
        healthActivityTypeToHealthkit[.HIKING] = .hiking
        healthActivityTypeToHealthkit[.HUNTING] = .hunting
        healthActivityTypeToHealthkit[.PLAY] = .play
        healthActivityTypeToHealthkit[.CROSS_COUNTRY_SKIING] = .crossCountrySkiing
        healthActivityTypeToHealthkit[.CURLING] = .curling
        healthActivityTypeToHealthkit[.DOWNHILL_SKIING] = .downhillSkiing
        healthActivityTypeToHealthkit[.SNOW_SPORTS] = .snowSports
        healthActivityTypeToHealthkit[.SNOWBOARDING] = .snowboarding
        healthActivityTypeToHealthkit[.SKATING] = .skatingSports
        healthActivityTypeToHealthkit[.SKATING_CROSS] = .skatingSports // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.SKATING_INDOOR] = .skatingSports // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.SKATING_INLINE] = .skatingSports // Supported due to combining with Android naming
        healthActivityTypeToHealthkit[.PADDLE_SPORTS] = .paddleSports
        healthActivityTypeToHealthkit[.ROWING] = .rowing
        healthActivityTypeToHealthkit[.SAILING] = .sailing
        healthActivityTypeToHealthkit[.SURFING_SPORTS] = .surfingSports
        healthActivityTypeToHealthkit[.SWIMMING] = .swimming
        healthActivityTypeToHealthkit[.WATER_FITNESS] = .waterFitness
        healthActivityTypeToHealthkit[.WATER_POLO] = .waterPolo
        healthActivityTypeToHealthkit[.WATER_SPORTS] = .waterSports
        healthActivityTypeToHealthkit[.BOXING] = .boxing
        healthActivityTypeToHealthkit[.KICKBOXING] = .kickboxing
        healthActivityTypeToHealthkit[.MARTIAL_ARTS] = .martialArts
        healthActivityTypeToHealthkit[.TAI_CHI] = .taiChi
        healthActivityTypeToHealthkit[.WRESTLING] = .wrestling
        healthActivityTypeToHealthkit[.OTHER] = .other

        healthkitTypeToHealth[.archery] = .ARCHERY
        healthkitTypeToHealth[.bowling] = .BOWLING
        healthkitTypeToHealth[.fencing] = .FENCING
        healthkitTypeToHealth[.gymnastics] = .GYMNASTICS
        healthkitTypeToHealth[.trackAndField] = .TRACK_AND_FIELD
        healthkitTypeToHealth[.americanFootball] = .AMERICAN_FOOTBALL
        healthkitTypeToHealth[.australianFootball] = .AUSTRALIAN_FOOTBALL
        healthkitTypeToHealth[.baseball] = .BASEBALL
        healthkitTypeToHealth[.basketball] = .BASKETBALL
        healthkitTypeToHealth[.cricket] = .CRICKET
        healthkitTypeToHealth[.discSports] = .DISC_SPORTS
        healthkitTypeToHealth[.handball] = .HANDBALL
        healthkitTypeToHealth[.hockey] = .HOCKEY
        healthkitTypeToHealth[.lacrosse] = .LACROSSE
        healthkitTypeToHealth[.rugby] = .RUGBY
        healthkitTypeToHealth[.soccer] = .SOCCER
        healthkitTypeToHealth[.softball] = .SOFTBALL
        healthkitTypeToHealth[.volleyball] = .VOLLEYBALL
        healthkitTypeToHealth[.preparationAndRecovery] = .PREPARATION_AND_RECOVERY
        healthkitTypeToHealth[.flexibility] = .FLEXIBILITY
        healthkitTypeToHealth[.walking] = .WALKING
        healthkitTypeToHealth[.running] = .RUNNING
        healthkitTypeToHealth[.wheelchairWalkPace] = .WHEELCHAIR_WALK_PACE
        healthkitTypeToHealth[.wheelchairRunPace] = .WHEELCHAIR_RUN_PACE
        healthkitTypeToHealth[.cycling] = .BIKING
        healthkitTypeToHealth[.handCycling] = .HAND_CYCLING
        healthkitTypeToHealth[.coreTraining] = .CORE_TRAINING
        healthkitTypeToHealth[.elliptical] = .ELLIPTICAL
        healthkitTypeToHealth[.functionalStrengthTraining] = .FUNCTIONAL_STRENGTH_TRAINING
        healthkitTypeToHealth[.traditionalStrengthTraining] = .TRADITIONAL_STRENGTH_TRAINING
        healthkitTypeToHealth[.crossTraining] = .CROSS_TRAINING
        healthkitTypeToHealth[.mixedCardio] = .MIXED_CARDIO
        healthkitTypeToHealth[.highIntensityIntervalTraining] = .HIGH_INTENSITY_INTERVAL_TRAINING
        healthkitTypeToHealth[.jumpRope] = .JUMP_ROPE
        healthkitTypeToHealth[.stairClimbing] = .STAIR_CLIMBING
        healthkitTypeToHealth[.stairs] = .STAIRS
        healthkitTypeToHealth[.stepTraining] = .STEP_TRAINING
        healthkitTypeToHealth[.fitnessGaming] = .FITNESS_GAMING
        healthkitTypeToHealth[.barre] = .BARRE
        healthkitTypeToHealth[.yoga] = .YOGA
        healthkitTypeToHealth[.mindAndBody] = .MIND_AND_BODY
        healthkitTypeToHealth[.pilates] = .PILATES
        healthkitTypeToHealth[.badminton] = .BADMINTON
        healthkitTypeToHealth[.racquetball] = .RACQUETBALL
        healthkitTypeToHealth[.squash] = .SQUASH
        healthkitTypeToHealth[.tableTennis] = .TABLE_TENNIS
        healthkitTypeToHealth[.tennis] = .TENNIS
        healthkitTypeToHealth[.climbing] = .CLIMBING
        healthkitTypeToHealth[.equestrianSports] = .EQUESTRIAN_SPORTS
        healthkitTypeToHealth[.fishing] = .FISHING
        healthkitTypeToHealth[.golf] = .GOLF
        healthkitTypeToHealth[.hiking] = .HIKING
        healthkitTypeToHealth[.hunting] = .HUNTING
        healthkitTypeToHealth[.play] = .PLAY
        healthkitTypeToHealth[.crossCountrySkiing] = .CROSS_COUNTRY_SKIING
        healthkitTypeToHealth[.curling] = .CURLING
        healthkitTypeToHealth[.downhillSkiing] = .DOWNHILL_SKIING
        healthkitTypeToHealth[.snowSports] = .SNOW_SPORTS
        healthkitTypeToHealth[.snowboarding] = .SNOWBOARDING
        healthkitTypeToHealth[.skatingSports] = .SKATING
        healthkitTypeToHealth[.paddleSports] = .PADDLE_SPORTS
        healthkitTypeToHealth[.rowing] = .ROWING
        healthkitTypeToHealth[.sailing] = .SAILING
        healthkitTypeToHealth[.surfingSports] = .SURFING_SPORTS
        healthkitTypeToHealth[.swimming] = .SWIMMING
        healthkitTypeToHealth[.waterFitness] = .WATER_FITNESS
        healthkitTypeToHealth[.waterPolo] = .WATER_POLO
        healthkitTypeToHealth[.waterSports] = .WATER_SPORTS
        healthkitTypeToHealth[.boxing] = .BOXING
        healthkitTypeToHealth[.kickboxing] = .KICKBOXING
        healthkitTypeToHealth[.martialArts] = .MARTIAL_ARTS
        healthkitTypeToHealth[.taiChi] = .TAI_CHI
        healthkitTypeToHealth[.wrestling] = .WRESTLING
        healthkitTypeToHealth[.other] = .OTHER

      if #available(iOS 14.0, *) {
          healthActivityTypeToHealthkit[.CARDIO_DANCE] = HKWorkoutActivityType.cardioDance
          healthActivityTypeToHealthkit[.SOCIAL_DANCE] = HKWorkoutActivityType.socialDance
          healthActivityTypeToHealthkit[.PICKLEBALL] = HKWorkoutActivityType.pickleball
          healthActivityTypeToHealthkit[.COOLDOWN] = HKWorkoutActivityType.cooldown

          healthkitTypeToHealth[.cardioDance] = .CARDIO_DANCE
          healthkitTypeToHealth[.socialDance] = .SOCIAL_DANCE
          healthkitTypeToHealth[.pickleball] = .PICKLEBALL
          healthkitTypeToHealth[.cooldown] = .COOLDOWN
      }
    }


    func toHealthkitActivityType(healthActivityType: String) -> HKWorkoutActivityType? {
        return healthActivityTypeToHealthkit[HealthWorkoutActivityType.withLabel(healthActivityType) ?? .OTHER];
    }

    func toHealthActivityType(hkActivityType: HKWorkoutActivityType) -> HealthWorkoutActivityType {
        return healthkitTypeToHealth[hkActivityType] ?? .OTHER;
    }
}